Services/MessageService.cs

```diff
@@ -44,15 +56,77 @@ public IEnumerable<Message> GetByLimit(string groupId, int limit, string lastMes
+            var message = new Message(content: content, fromId: int.Parse(userId), groupId: groupId, dateSent: DateTime.UtcNow, status: 0);
             _messages.Add(message);  
         }
 
         public void SaveMessages(string groupId)
         {
             _messageRepository.Add(_messages, groupId);
         }
+
+        public void CreateConversation(int userOneId, int userTwoId, int listingId)
+        {
+            var groupId = Guid.NewGuid().ToString("N");
+            _messageRepository.CreateConversation(userOneId, userTwoId, listingId, groupId);
+        }
+
+        public IEnumerable<ConversationDto> GetConversations(int userId)
+        {
+            var conversations = _messageRepository.GetConversations(userId);
+            var listings = _listingRepository.GetListingsById(
+                (from convo in conversations
+                select convo.ListingId)
+                .ToList());
+            var messages = new List<Message>();
+            foreach (var convo in conversations)
+            {
+                var message = _messageRepository.GetByLimit(convo.GroupId, 1);
+                messages.AddRange(message);
+            }
+            var query = from c in conversations
+                        join l in listings
+                        on c.ListingId equals l.Id
+                        select new  {
+                                    ListingName = l.Name,
+                                    ListingImg = l.ImagePath,
+                                    GroupId = c.GroupId,
+                                    ListingId = l.Id,
+                                    UserOneId = c.UserOneId,
+                                    UserTwoId = c.UserTwoId,
+                                    };
+            var result = (from q in query
+                        select new ConversationDto {
+                        ListingName = q.ListingName,
+                        LastSender = "",
+                        LastMessage = "",
+                        UserOneId = q.UserOneId,
+                        UserTwoId = q.UserTwoId,
+                        UserOneName = _userService.GetUserById(q.UserOneId).FirstName,
+                        UserTwoName = _userService.GetUserById(q.UserTwoId).FirstName,
+                        ListingImg = q.ListingImg,
+                        GroupId = q.GroupId,
+                        ListingId = q.ListingId
+                        }).ToList();
+            
+            foreach (var item in result)
+            {
+                var msg = messages.FirstOrDefault(x => x.GroupId == item.GroupId);
+                if(msg != null)
+                {
+                    var user = _userService.GetUserById(msg.FromId);
+                    item.LastSender = user.FirstName;
+                    item.LastMessage = msg.Content;
+                    item.DateSent = msg.DateSent;
+                }
+
+            }
```

Comment:
I believe this foreach could be combined with foreach in line 82 and it would remove the need to run message.FirstOrDefault.

Reply:
Combined both loops but FirstOrDefault() is still needed since GetByLimit() returns an IEnumerable.